core/src/main/java/com/nisovin/magicspells/util/handlers/RecipeHandler.java

```diff
@@ -0,0 +1,208 @@
+package com.nisovin.magicspells.util.handlers;
+
+import java.util.Set;
+import java.util.List;
+import java.util.HashSet;
+
+import org.bukkit.Bukkit;
+import org.bukkit.Material;
+import org.bukkit.inventory.*;
+import org.bukkit.NamespacedKey;
+import org.bukkit.entity.Player;
+import org.bukkit.event.Listener;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.player.PlayerJoinEvent;
+import org.bukkit.configuration.ConfigurationSection;
+
+import com.nisovin.magicspells.util.Util;
+import com.nisovin.magicspells.MagicSpells;
+
+public class RecipeHandler implements Listener {
+
+    private static Set<NamespacedKey> namespaces = new HashSet<>();
+
+    public RecipeHandler() {
+        Bukkit.getPluginManager().registerEvents(this, MagicSpells.plugin);
+    }
+
+    public static void create(ConfigurationSection config) {
+        String type = config.getString("type", "shaped");
+        if (type == null) {
+            MagicSpells.error("Recipe '" + config.getName() + "' has an invalid 'type' defined.");
+            return;
+        }
+        type = type.toLowerCase();
+
+        ItemStack result = getItem(config.getString("result"), "Recipe '" + config.getName() + "' has an invalid 'result' item defined.");
+        if (result == null) return;
+
+        int quantity = config.getInt("quantity", 1);
+        if (quantity > 1) result.setAmount(quantity);
+
+        String group = config.getString("group");
+        if (group == null) group = "";
+
+        String namespaceKey = config.getString("namespace");
+        if (namespaceKey == null || namespaceKey.isEmpty()) {
+            MagicSpells.error("Recipe '" + config.getName() + "' has an invalid 'namespace' defined.");
+            return;
+        }
+        NamespacedKey namespace;
+        try {
+            namespace = new NamespacedKey(MagicSpells.plugin, namespaceKey.toLowerCase());
+        }
+        catch (IllegalArgumentException e) {
+            MagicSpells.error("Recipe '" + config.getName() + "' has an invalid 'namespace' defined: " + namespaceKey);
+            MagicSpells.handleException(e);
+            return;
+        }
```

Comment:
I'd **strongly** encourage reconsideration of the use of namespaceKey and namespace as variable names. They seem to be rather misleading as is.